src/include/network/connection_handle.h

```diff
@@ -0,0 +1,203 @@
+   * client query.
+   */
+  void StopReceivingNetworkEvent() {
+    EventUtil::EventDel(network_event_);
+
+    // TODO(tanujnay112) remove
+    const char *sample = "I have received your request.\n";
+    io_wrapper_->GetWriteQueue()->BufferWriteRaw(sample, strlen(sample) + 1);
+    io_wrapper_->GetWriteQueue()->ForceFlush();
+    event_active(workpool_event_, EV_WRITE, 0);
+  }
+
+ private:
+  /**
+   * A state machine is defined to be a set of states, a set of symbols it
+   * supports, and a function mapping each
+   * state and symbol pair to the state it should transition to. i.e.
+   * transition_graph = state * symbol -> state
+   *
+   * In addition to the transition system, our network state machine also needs
+   * to perform actions. Actions are
+   * defined as functions (lambdas, or closures, in various other languages) and
+   * is promised to be invoked by the
+   * state machine after each transition if registered in the transition graph.
+   *
+   * So the transition graph overall has type transition_graph = state * symbol
+   * -> state * action
+   */
+  class StateMachine {
+   public:
+    using action = Transition (*)(ConnectionHandle &);
+    using transition_result = std::pair<ConnState, action>;
+    /**
+     * Runs the internal state machine, starting from the symbol given, until no
+     * more
+     * symbols are available.
+     *
+     * Each state of the state machine defines a map from a transition symbol to
+     * an action
+     * and the next state it should go to. The actions can either generate the
+     * next symbol,
+     * which means the state machine will continue to run on the generated
+     * symbol, or signal
+     * that there is no more symbols that can be generated, at which point the
+     * state machine
+     * will stop running and return, waiting for an external event (user
+     * interaction, or system event)
+     * to generate the next symbol.
+     *
+     * @param action starting symbol
+     * @param connection the network connection object to apply actions to
+     */
+    void Accept(Transition action, ConnectionHandle &connection);  // NOLINT
+    // clang-tidy is suppressed here as it complains about the reference param having
+    // no const qualifier as this must be casted into a (void*) to pass as an argument to METHOD_AS_CALLBACK
+    // in the forward dependencies of Accept in the state_machine
+
+   private:
+    /**
+     * delta is the transition function that defines, for each state, its
+     * behavior and the
+     * next state it should go to.
+     */
+    static transition_result Delta_(ConnState state, Transition transition);
+    ConnState current_state_ = ConnState::READ;
+  };
+
+  friend class StateMachine;
+  friend class ConnectionHandleFactory;
+
+  // A raw pointer is used here because references cannot be rebound.
+  ConnectionHandlerTask *conn_handler_;
+  std::unique_ptr<NetworkIoWrapper> io_wrapper_;
```

Comment:
Should this be unique_ptr of type PosixSocketIoWrapper?

Reply:
I think the point of this was to abstract away the implementation of NetworkIoWrapper in order for ease of switching out different implementations in Peloton?

Reply:
It seems unlikely to me that we'll be implementing other network wrappers. So two options:
1. Change it to Posix... wrapper, and add a comment that this should be changed to be a generic NetworkIoWrapper if switchable wrappers implemented in future.
2. Add a comment stating that this is currently a PosixSocketIOWrapper. 
I'd probably do 1), because it will make looking up classes and definitions in CLion work naturally, rather than having to search and find the actual class used.

Reply:
Alright sounds good.